mender-update/http_resumer/http_resumer_test.cpp

```diff
@@ -861,3 +861,217 @@ TEST_F(DownloadResumerTest, SmallIntervalsErrorOnFirstRead) {
+
+	EXPECT_TRUE(body_handler_called);
+}
+
+TEST_F(DownloadResumerTest, UserCancelInHeaderHandler) {
+	TestEventLoop loop(chrono::seconds(10));
+
+	// Server
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+		},
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+
+			auto result = exp_req.value()->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetHeader("Content-Length", to_string(RangeBodyOfXes::TARGET_BODY_SIZE));
+			resp->SetBodyReader(make_shared<RangeBodyOfXes>());
+			resp->SetStatusCodeAndMessage(200, "Success");
+			resp->AsyncReply([](error::Error err) { ASSERT_NE(error::NoError, err); });
+		});
+
+	// Request
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	req->SetAddress("http://127.0.0.1:" TEST_PORT);
+
+	// Client
+	http::ClientConfig client_config;
+	shared_ptr<http_resumer::DownloadResumerClient> client =
+		make_shared<http_resumer::DownloadResumerClient>(client_config, loop);
+	client->SetSmallestWaitInterval(chrono::seconds(2));
+
+	client->AsyncCall(
+		req,
+		[&client](http::ExpectedIncomingResponsePtr exp_resp) { client->Cancel(); },
+		[](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_FALSE(exp_resp);
+			EXPECT_EQ(exp_resp.error().code, make_error_condition(errc::operation_canceled));
+		});
+
+	events::Timer timer(loop);
+	timer.AsyncWait(chrono::milliseconds(500), [&loop](error::Error err) { loop.Stop(); });
+
+	loop.Run();
+}
+
+TEST_F(DownloadResumerTest, UserCancelInBodyHandler) {
+	TestEventLoop loop(chrono::seconds(10));
+
+	// Server
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+		},
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+
+			auto result = exp_req.value()->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetStatusCodeAndMessage(200, "Success");
+			resp->AsyncReply([](error::Error err) { ASSERT_EQ(error::NoError, err); });
+		});
+
+	// Request
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	req->SetAddress("http://127.0.0.1:" TEST_PORT);
+
+	// Client
+	http::ClientConfig client_config;
+	shared_ptr<http_resumer::DownloadResumerClient> client =
+		make_shared<http_resumer::DownloadResumerClient>(client_config, loop);
+	client->SetSmallestWaitInterval(chrono::seconds(2));
+
+	client->AsyncCall(
+		req,
+		[](http::ExpectedIncomingResponsePtr exp_resp) {},
+		[&client](http::ExpectedIncomingResponsePtr exp_resp) { client->Cancel(); });
```

Comment:
It should already be cancelled ("finished" is perhaps a better word), once the body handler is called, so this is, in theory, a no-op.

Reply:
To be honest, I copied these two tests almost verbatim from [http_test.cpp](https://github.com/mendersoftware/mender/blob/b6b72ae7f0ccfa90ac1ed9bbf2bd572aab8d1401/common/http_test.cpp#L929).

I see that this test doesn't do much - I repurpose to do stream actual data and cancel half way from a timer.

Reply:
This was quite useful. It revealed an interesting bug: the resumer was _unstoppable_! After a `Cancel()` it would just interpret the call to body handler (with operation cancelled error) as a generic error and resume again :laughing: 

Fixed now in the push coming up.

Reply:
Haha, nice! The unstoppable resumer! :muscle: